In [1]:
import pandas as pd
import numpy as np
import math
#neural networks, support vector machines, random forests anda KNN classifier. 
from sklearn.ensemble import RandomForestClassifier
from sklearn.datasets import make_classification
from sklearn import svm
from sklearn.metrics import accuracy_score
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split
from sklearn.neural_network import MLPClassifier
from sklearn.naive_bayes import GaussianNB

- The response variable is y in column 179
- All subjects falling in classes 2, 3, 4, and 5 are subjects who did not have epileptic seizure
- 1- Recording of seizure activity
- The Explanatory variables X1, X2, ..., X178

In [2]:
df = pd.read_csv("data.csv")
df

Unnamed: 0   X1   X2   X3   X4   X5   X6   X7   X8   X9  ...  X170  \
0      X21.V1.791  135  190  229  223  192  125   55   -9  -33  ...   -17   
1      X15.V1.924  386  382  356  331  320  315  307  272  244  ...   164   
2         X8.V1.1  -32  -39  -47  -37  -32  -36  -57  -73  -85  ...    57   
3       X16.V1.60 -105 -101  -96  -92  -89  -95 -102 -100  -87  ...   -82   
4       X20.V1.54   -9  -65  -98 -102  -78  -48  -16    0  -21  ...     4   
...           ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   ...   
11495  X22.V1.114  -22  -22  -23  -26  -36  -42  -45  -42  -45  ...    15   
11496  X19.V1.354  -47  -11   28   77  141  211  246  240  193  ...   -65   
11497    X8.V1.28   14    6  -13  -16   10   26   27   -9    4  ...   -65   
11498  X10.V1.932  -40  -25   -9  -12   -2   12    7   19   22  ...   121   
11499  X16.V1.210   29   41   57   72   74   62   54   43   31  ...   -59   

       X171  X172  X173  X174  X175  X176  X177  X178  y  
0       -15   -31   -77  -103  -127  -116   -83   -51  4  
1       150   146   152   157   156   154   143   129  1  
2        64    48    19   -12   -30   -35   -35   -36  5  
3       -81   -80   -77   -85   -77   -72   -69   -65  5  
4         2   -12   -32   -41   -65   -83   -89   -73  5  
...     ...   ...   ...   ...   ...   ...   ...   ... ..  
11495    16    12     5    -1   -18   -37   -47   -48  2  
11496   -33    -7    14    27    48    77   117   170  1  
11497   -48   -61   -62   -67   -30    -2    -1    -8  5  
11498   135   148   143   116    86    68    59    55  3  
11499   -25    -4     2     5     4    -2     2    20  4  

[11500 rows x 180 columns]

In [3]:
df = df.drop(columns=['Unnamed: 0'])
df.loc[df['y'] > 1, 'y'] = 0
df

X1   X2   X3   X4   X5   X6   X7   X8   X9  X10  ...  X170  X171  \
0      135  190  229  223  192  125   55   -9  -33  -38  ...   -17   -15   
1      386  382  356  331  320  315  307  272  244  232  ...   164   150   
2      -32  -39  -47  -37  -32  -36  -57  -73  -85  -94  ...    57    64   
3     -105 -101  -96  -92  -89  -95 -102 -100  -87  -79  ...   -82   -81   
4       -9  -65  -98 -102  -78  -48  -16    0  -21  -59  ...     4     2   
...    ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   ...   ...   
11495  -22  -22  -23  -26  -36  -42  -45  -42  -45  -49  ...    15    16   
11496  -47  -11   28   77  141  211  246  240  193  136  ...   -65   -33   
11497   14    6  -13  -16   10   26   27   -9    4   14  ...   -65   -48   
11498  -40  -25   -9  -12   -2   12    7   19   22   29  ...   121   135   
11499   29   41   57   72   74   62   54   43   31   23  ...   -59   -25   

       X172  X173  X174  X175  X176  X177  X178  y  
0       -31   -77  -103  -127  -116   -83   -51  0  
1       146   152   157   156   154   143   129  1  
2        48    19   -12   -30   -35   -35   -36  0  
3       -80   -77   -85   -77   -72   -69   -65  0  
4       -12   -32   -41   -65   -83   -89   -73  0  
...     ...   ...   ...   ...   ...   ...   ... ..  
11495    12     5    -1   -18   -37   -47   -48  0  
11496    -7    14    27    48    77   117   170  1  
11497   -61   -62   -67   -30    -2    -1    -8  0  
11498   148   143   116    86    68    59    55  0  
11499    -4     2     5     4    -2     2    20  0  

[11500 rows x 179 columns]

In [4]:
y = df['y']
x = df
x = x.drop(columns=['y'])

In [5]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.25, random_state=0)

In [18]:
rfcl = RandomForestClassifier(max_depth=5, random_state=1)
rfcl.fit(x_train, y_train)
rfcl.score(x_test,y_test)

0.9495652173913044

Random Forest Classifier has an accuracy of 94.96%

In [7]:
# clf = svm.SVC(kernel='precomputed')
clf = svm.SVC(kernel='rbf')
clf.fit(x_train, y_train)
clf.score(x_test, y_test)

0.976

Support Vector Machine Classifier has an accuracy of 97.6% 

In [27]:
nncl = MLPClassifier(solver='lbfgs', hidden_layer_sizes=(10,), random_state=1)
nncl.fit(x_train, y_train)
nncl.score(x_test, y_test)

/Users/georgebriggs/opt/anaconda3/lib/python3.9/site-packages/sklearn/neural_network/_multilayer_perceptron.py:500: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
  self.n_iter_ = _check_optimize_result("lbfgs", opt_res, self.max_iter)


0.9634782608695652

Neural network has an accuracy of 96.21%

In [12]:
neigh = KNeighborsClassifier(n_neighbors=3)
neigh.fit(x_train, y_train)
neigh.score(x_test, y_test)

0.928695652173913

KNN has an accuracy of 92.87

In [13]:
gnb = GaussianNB()
gnb.fit(x_train, y_train)
gnb.score(x_test, y_test)

0.959304347826087

In [51]:
# y_test[4313]

In [19]:
correct = 0
total = 0
predictions = []
for i in range(len(x_test)):
    p1 = rfcl.predict(x_test.iloc[[i]])
    p2 = nncl.predict(x_test.iloc[[i]])
    p3 = neigh.predict(x_test.iloc[[i]])
    p4 = clf.predict(x_test.iloc[[i]])
    p5 = gnb.predict(x_test.iloc[[i]])
    predictions.append([p1,p2,p3,p4,p5])
#     if (p1[0]+p2[0]+p3[0]+p4[0])>=2:
#         predict=1
#     else:
#         predict=0
#     predictions.append(predict)


In [29]:
actual = []
for element in y_test:
    actual.append(element)


In [56]:
print(len(actual), len(predictions))

2875 2875


In [50]:
correct = 0
#type 1 is flase-positive
type_1 = 0
type_2 = 0
for i in range(len(actual)):
    s = predictions[i][1]+predictions[i][3]+predictions[i][4] +predictions[i][1]#3 best algorithms
#     s = max(predictions[i]) #any positive predictions
#     s = sum(predictions[i]) #all values
#     s = predictions[i][3] #most accurate model
    if s>0:
        prediction = 1
    else:
        prediction = 0
    if actual[i]==prediction:
        correct += 1
    else:
#         print(s, actual[i])
        if actual[i]==1 and prediction==0:
            type_2 += 1
        else:
            type_1 += 1
print(type_1, type_2)

print("Accuracy:", float(correct/len(actual)))
print("Type 2 errors:", float(type_2/len(actual)))


65 28
Accuracy: 0.9676521739130435
Type 2 errors: 0.009739130434782608
